bsdsockets: Add proxy socket support

TSRBerry opened the pull request:
This PR is part of our ongoing effort to upstream our LDN implementation (#4716).

The new `ManagedProxySocket` allows SOCKS5 compatible proxy servers to intercept and manipulate traffic from the guest.

I implemented many parts of this according to [RFC1928 - SOCKS5 Protocol](https://datatracker.ietf.org/doc/html/rfc1928), but since there are some things I didn't do, this isn't a compliant implementation. I added a TODO comment for everything that needs to be implemented to be SOCKS5 compliant.

I tested the TCP connect part of this implementation using @Raytwo's Cobalt mod which connects to the GitHub API for a version check and mitmproxy. Sadly mitmproxy does not support UDP for SOCKS5, so this part still needs to be tested.

There are a few other things I'm not sure about where feedback and/or testing would be nice:
- [ ] Does proxying UDP packets actually work?
- [ ] Does the TCP proxy work in server mode (using the `Bind` command)?
- [ ] For `LocalEndPoint` and `RemoteEndPoint` I could use the endpoints I receive from the proxy server like I do now, or use the actual endpoints from the sockets directly. Which one is better? (Both could work, don't consider RyuLDN for this question, since it doesn't matter for the way I want to add RyuLDN with this proxy implementation)
- [ ] Should I throw exceptions for proxy responses containing errors or should I convert them to `LinuxError`s somehow?
- [ ] Is the way the `Handle` property works fine?
- [ ] Should I provide an implementation for `RecvMMsg` and `SendMMsg`?

riperiperi commented:
>  Does proxying UDP packets actually work?

Why is this an unknown? This is what nearly all LDN games require.

TSRBerry commented:
I just rebased this PR and replaced the old SOCKS5 implementation with [RyuSOCKS](https://www.nuget.org/packages/RyuSocks)!

I also adjusted the description accordingly and listed all the things that need to be done, before I mark this PR as ready to review.

I expect to finish this todo list within the next few weeks.

TSRBerry commented:
Seems like the stream wrapper in `RyuSocks.SocksClientStream` isn't working correctly yet.

I just repeated the mitmproxy test with Cobalt and successfully froze Ryujinx before the request even reached the proxy server.

Once I figure out how to fix that, I'll be able to mark this PR as ready to review again.

TSRBerry commented:
Finally ready for review! :D

gdkchan commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocketPollManager.cs:
```diff
@@ -0,0 +1,211 @@
+    {
+        private static ManagedProxySocketPollManager _instance;
+
+        public static ManagedProxySocketPollManager Instance
```
I'm not a fan of those static instances, but it seems the other poll managers do the same, so it should be fine for now.

TSRBerry replied:
I suppose that's something we could tackle once we move this to `Ryujinx.Horizon` as well?

gdkchan replied:
Yes.

gdkchan reviewed (approved):
lgtm, thanks. Should wait for @riperiperi review before it is merged, I did not work on LDN so I'm not the best person to review this.

riperiperi commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Proxy/ProxyManager.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    public static class ProxyManager
+    {
+        private static readonly Dictionary<string, EndPoint> _proxyEndpoints = new();
```
Maybe this dictionary should be thread safe? I can imagine other threads adding Endpoints and maybe multiple threads making clients at once, though I don't think the BSD service can do that right now.

riperiperi commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocket.cs:
```diff
@@ -0,0 +1,472 @@
+            {
+                EndPoint temp = new IPEndPoint(IPAddress.Any, 0);
+
+                receiveSize = ProxyClient.ReceiveFrom(proxyBuffer, WinSockHelper.ConvertBsdSocketFlags(flags), ref temp);
```
Is this missing a copy to `buffer`? Its data is never updated.

TSRBerry replied:
Yes it is, thank you.

riperiperi commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocket.cs:
```diff
@@ -0,0 +1,472 @@
+                    out SocketError errorCode
+                );
+
+                proxyBuffer[..receiveSize].CopyTo(buffer);
```
Is the wrapper at the end of the buffer? If it's at the start, this should be `proxyBuffer[^recieveSize..]`.

TSRBerry replied:
Currently it will be at the start, but this is up to the implementation of the authentication method and the proxy command. The `Receive()` call already performs the unwrapping and copies everything to the start of the buffer.
I only use `[..receiveSize]` here so it doesn't attempt to copy more data than necessary.

riperiperi commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocket.cs:
```diff
@@ -0,0 +1,472 @@
+
+        public LinuxError Connect(IPEndPoint remoteEndPoint)
+        {
+            ProxyClient.RequestCommand = ProxyCommand.Connect;
```
Why do these commands need to be set on a property if a method specific to them is called later? (like `ProxyClient.Connect`)

TSRBerry replied:
Yeah, this part feels a little weird.

I was looking for a way to generate these command specific methods to avoid using a property for this, but as far as I know this would prevent people from adding/using their own commands and authentication methods, since the NuGet package only ships as a DLL and my code can't be modified anymore at that point.

Since I want people to be able to add commands and authentication methods to RyuSocks easily, I currently work around this problem by providing methods similar to `System.Net.Sockets.Socket` in `SocksClient` while still allowing them to be used with any provided proxy command.

I don't like how it currently works and I'm looking to improve it. Maybe I should provide a more general method that handles the proxy commands and remove the more specific ones?

Although I'd love to have something that could allow me to ship the RyuSOCKS NuGet package as source code and generate the methods with a source generator - in such a way that people could add an attribute or something to their own commands, so they also get picked up and handled by the source generator.

riperiperi commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocket.cs:
```diff
@@ -0,0 +1,472 @@
+
+        public void Disconnect()
+        {
+            ProxyClient.Disconnect();
```
Should this (and connection errors) reset `ProxyClient.RequestCommand`? I can't remember if games end up reusing sockets after they disconnect, but in case it's possible it might need reset to prevent errors on some of the methods.

TSRBerry replied:
Oh yeah, that's definitely safer. I'll do that.

riperiperi reviewed (commented):
Looks good, I have a few questions about certain parts of the code. I don't know if this is supposed to be testable in any form right now, I guess I can see if LAN and ldn_mitm still work properly.

TSRBerry commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocketPollManager.cs:
```diff
@@ -0,0 +1,211 @@
+                int actualTimeoutMicroseconds = timeoutMilliseconds == -1 ? -1 : timeoutMilliseconds * 1000;
+                int totalEvents = eventDict[SelectMode.SelectRead].Count + eventDict[SelectMode.SelectWrite].Count + eventDict[SelectMode.SelectError].Count;
+                // TODO: Maybe check all events first, wait for the timeout and then check the failed ones again?
+                int timeoutMicrosecondsPerEvent = actualTimeoutMicroseconds == -1 ? -1 : actualTimeoutMicroseconds / totalEvents;
```
The only way I can think of to fix this would be to provide a `Socket.Select()` method from `RyuSOCKS`, which then calls `Socket.Select()` with the underlying sockets itself?

If you think this solution could work I'd implement it for this PR. Otherwise or if you don't think it's important I'll open an issue for this at the `RyuSOCKS` repo and work on that some time after completing LDN upstreaming.

TSRBerry commented on src/Ryujinx.HLE/HOS/Services/Sockets/Bsd/Impl/ManagedProxySocketPollManager.cs:
```diff
@@ -0,0 +1,211 @@
+                {
+                    outputEvents |= PollEventTypeMask.Error;
+
+                    // TODO: Check ProxyClient.Connected and ProxyClient.IsBound when implemented.
```
Forgot about that - it's implemented now!
